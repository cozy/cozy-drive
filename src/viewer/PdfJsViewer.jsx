import React, { Component } from 'react'
import { Document, Page } from 'react-pdf/dist/entry.webpack'
import cx from 'classnames'
import throttle from 'lodash/throttle'
import { Button, Spinner } from 'cozy-ui/react'
import withFileUrl from './withFileUrl'
import NoViewer from './NoViewer'
import styles from './styles'

export const MIN_SCALE = 0.25
export const MAX_SCALE = 3

const ToolbarButton = ({ icon, onClick, disabled }) => (
  <Button
    iconOnly
    subtle
    theme="secondary"
    className="u-p-half u-m-half"
    icon={icon}
    onClick={onClick}
    disabled={disabled}
  />
)

export class PdfJsViewer extends Component {
  state = {
    totalPages: 1,
    scale: 1,
    currentPage: 1,
    loaded: false,
    errored: false,
    width: null
  }

  constructor(props) {
    super(props)

    this.wrapper = null
    this.resizeListener = null

    this.onLoadSuccess = this.onLoadSuccess.bind(this)
    this.onLoadError = this.onLoadError.bind(this)
    this.nextPage = this.nextPage.bind(this)
    this.previousPage = this.previousPage.bind(this)
    this.scaleUp = this.scaleUp.bind(this)
    this.scaleDown = this.scaleDown.bind(this)
    this.setWrapperSize = this.setWrapperSize.bind(this)
  }

  componentDidMount() {
    this.setWrapperSize()
    this.resizeListener = throttle(this.setWrapperSize, 500)
    window.addEventListener('resize', this.resizeListener)
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.resizeListener)
  }

  setWrapperSize = () => {
    this.setState({ width: this.wrapper.getBoundingClientRect().width })
  }

  async onLoadSuccess({ numPages }) {
    this.setState({
      totalPages: numPages,
      loaded: true
    })
  }

  async onLoadError(error) {
    console.warn(error)
    this.setState({
      errored: true
    })
  }

  nextPage() {
    this.setState(state => ({
      currentPage: Math.min(state.currentPage + 1, state.totalPages)
    }))
  }

  previousPage() {
    this.setState(state => ({
      currentPage: Math.max(state.currentPage - 1, 1)
    }))
  }

  scaleUp() {
    this.setState(state => ({
      scale: Math.min(state.scale + 0.25, MAX_SCALE)
    }))
  }

  scaleDown() {
    this.setState(state => ({
      scale: Math.max(state.scale - 0.25, MIN_SCALE)
    }))
  }

  render() {
    const { url } = this.props
    const {
      loaded,
      errored,
      totalPages,
      currentPage,
      scale,
      width
    } = this.state

    if (errored) return <NoViewer file={url} />
    const pageWidth = width ? width * scale : null // newer versions of react-pdf do that automatically

    return (
      <div
        data-test-id="viewer-pdf"
        className={styles['pho-viewer-pdfviewer']}
        ref={ref => (this.wrapper = ref)}
      >
        <Document
          file={url}
          onLoadSuccess={this.onLoadSuccess}
          onLoadError={this.onLoadError}
          className={styles['pho-viewer-pdfviewer-pdf']}
          loading={<Spinner size="xxlarge" middle noMargin color="white" />}
        >
          <Page pageNumber={currentPage} width={pageWidth} />
        </Document>
        {loaded && (
          <div
            className={cx(styles['pho-viewer-pdfviewer-toolbar'], 'u-p-half')}
          >
            <span className="u-mh-half">
              <ToolbarButton
                icon="top"
                onClick={this.previousPage}
                disabled={currentPage === 1}
              />
              {currentPage}/{totalPages}
              <ToolbarButton
                icon="bottom"
                onClick={this.nextPage}
                disabled={currentPage === totalPages}
              />
            </span>
            <span className="u-mh-half">
              <ToolbarButton
                icon="dash"
                onClick={this.scaleDown}
                disabled={scale === MIN_SCALE}
              />
              <ToolbarButton
                icon="plus"
                onClick={this.scaleUp}
                disabled={scale === MAX_SCALE}
              />
            </span>
          </div>
        )}
      </div>
    )
  }
}

export default withFileUrl(PdfJsViewer)
