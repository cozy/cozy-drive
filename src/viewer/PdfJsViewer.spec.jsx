import { PdfJsViewer, MIN_SCALE, MAX_SCALE, MAX_PAGES } from './PdfJsViewer'
import React from 'react'
import { shallow } from 'enzyme'

describe('PDFViewer', () => {
  let component
  const panGestureMock = jest.fn()
  const swipeGestureMock = jest.fn()
  const gesturesMock = {
    get: jest.fn(type => ({
      set: type === 'pan' ? panGestureMock : swipeGestureMock
    }))
  }
  beforeEach(() => {
    component = shallow(<PdfJsViewer url={'test'} gestures={gesturesMock} />)
  })
  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('desture integration', () => {
    let instance
    beforeEach(() => {
      instance = component.instance()
    })

    it('should disable gestures when zooming in', () => {
      instance.scaleUp()
      expect(component.state('scale')).toBeGreaterThan(1)
      expect(panGestureMock).toHaveBeenCalledWith({ enable: false })
      expect(swipeGestureMock).toHaveBeenCalledWith({ enable: false })
      instance.scaleUp()
      expect(panGestureMock).toHaveBeenCalledTimes(1)
      expect(swipeGestureMock).toHaveBeenCalledTimes(1)
    })

    it('should leave gestures alone when zooming out', () => {
      instance.scaleDown()
      expect(component.state('scale')).toBeLessThan(1)
      expect(panGestureMock).not.toHaveBeenCalled()
      expect(swipeGestureMock).not.toHaveBeenCalled()
    })

    it('should re-enable gestures when zooming back out', () => {
      instance.scaleUp()
      expect(component.state('scale')).toBeGreaterThan(1)
      expect(panGestureMock).toHaveBeenCalledWith({ enable: false })
      expect(swipeGestureMock).toHaveBeenCalledWith({ enable: false })
      instance.scaleDown()
      expect(component.state('scale')).toBe(1)
      expect(panGestureMock).toHaveBeenCalledWith({ enable: true })
      expect(swipeGestureMock).toHaveBeenCalledWith({ enable: true })
    })
  })

  describe('with a valid PDF', () => {
    let instance
    beforeEach(() => {
      instance = component.instance()
      instance.onLoadSuccess({ numPages: 3 })
    })

    it('should start with default options', () => {
      expect(component.state('loaded')).toBe(true)
      expect(component.state('totalPages')).toBe(3)
      expect(component.state('currentPage')).toBe(1)
      expect(component.state('scale')).toBe(1)
    })

    it('should flip to the next page while possible', () => {
      instance.nextPage()
      expect(component.state('currentPage')).toBe(2)
      instance.nextPage()
      expect(component.state('currentPage')).toBe(3)
      instance.nextPage()
      expect(component.state('currentPage')).toBe(3)
    })

    it('should flip to the previous page while possible', () => {
      instance.nextPage()
      expect(component.state('currentPage')).toBe(2)
      instance.previousPage()
      expect(component.state('currentPage')).toBe(1)
      instance.previousPage()
      expect(component.state('currentPage')).toBe(1)
    })

    it('should scale up to a certain point', () => {
      const initialScale = component.state('scale')
      instance.scaleUp()
      expect(component.state('scale')).toBeGreaterThan(initialScale)

      for (let i = 0; i < 10; i++) instance.scaleUp()
      expect(component.state('scale')).toEqual(MAX_SCALE)
    })

    it('should scale down to a certain point', () => {
      const initialScale = component.state('scale')
      instance.scaleDown()
      expect(component.state('scale')).toBeLessThan(initialScale)

      for (let i = 0; i < 10; i++) instance.scaleDown()
      expect(component.state('scale')).toEqual(MIN_SCALE)
    })
  })

  describe('a PDF with few pages', () => {
    beforeEach(() => {
      component.instance().onLoadSuccess({ numPages: MAX_PAGES })
    })

    it('should render all the pages', () => {
      const pages = component.find('Page')
      expect(pages.length).toEqual(MAX_PAGES)
    })

    it('should not show pagination controls', () => {
      const pageUp = component.find({ icon: 'top' })
      const pageDown = component.find({ icon: 'bottom' })
      expect(pageUp.length).toEqual(0)
      expect(pageDown.length).toEqual(0)
    })
  })

  describe('a PDF with many pages', () => {
    beforeEach(() => {
      component.instance().onLoadSuccess({ numPages: MAX_PAGES + 1 })
    })

    it('should render only the current page', () => {
      const pages = component.find('Page')
      expect(pages.length).toEqual(1)
    })

    it('should show pagination controls', () => {
      const pageUp = component.find({ icon: 'top' })
      const pageDown = component.find({ icon: 'bottom' })
      expect(pageUp.length).toEqual(1)
      expect(pageDown.length).toEqual(1)
    })
  })

  describe('with a pdf that does not load', () => {
    it('should show a fallback', () => {
      component.instance().onLoadError('pdfviewer test error')
      expect(component.state('errored')).toBe(true)
      const wrapper = component.find('Wrapper')
      const noViewer = wrapper.dive().find('NoViewer')
      expect(noViewer.length).toBe(1)
    })
  })
})
