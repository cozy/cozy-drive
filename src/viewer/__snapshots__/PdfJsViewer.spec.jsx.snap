// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDFViewer with a pdf that does not load should show a fallback 1`] = `
<Wrapper
  file="test"
/>
`;
